StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c

```diff
@@ -0,0 +1,865 @@
+
+  @param[in]       PlatformHobList    Platform HOB list.
+  @param[in]       PlatformHobSize    Platform HOB size.
+  @param[in, out]  MemoryRegion       Memory regions.
+  @param[in, out]  MemoryRegionCount  Count of MMIO regions
+**/
+VOID
+CollectPlatformMemoryRegions (
+  IN UINT8                     *PlatformHobList,
+  IN UINTN                     PlatformHobSize,
+  IN OUT MM_IPL_MEMORY_REGION  *MemoryRegion,
+  IN OUT UINTN                 *MemoryRegionCount
+  )
+{
+  UINTN                 Index;
+  EFI_PEI_HOB_POINTERS  Hob;
+
+  ASSERT (MemoryRegionCount != NULL);
+  ASSERT (*MemoryRegionCount == 0 || MemoryRegion != NULL);
+
+  Index = 0;
+  //
+  // Get the HOB list for processing
+  //
+  Hob.Raw = PlatformHobList;
+
+  //
+  // Collect memory ranges
+  //
+  while (Hob.Raw < PlatformHobList + PlatformHobSize) {
+    if (Hob.Header->HobType == EFI_HOB_TYPE_RESOURCE_DESCRIPTOR) {
+      if (  (Hob.ResourceDescriptor->ResourceType == EFI_RESOURCE_MEMORY_MAPPED_IO)
+         || (Hob.ResourceDescriptor->ResourceType == EFI_RESOURCE_SYSTEM_MEMORY)
+         || (Hob.ResourceDescriptor->ResourceType == EFI_RESOURCE_FIRMWARE_DEVICE)
+         || (Hob.ResourceDescriptor->ResourceType == EFI_RESOURCE_MEMORY_RESERVED))
+      {
+        if (Index < *MemoryRegionCount) {
+          MemoryRegion[Index].Base   = Hob.ResourceDescriptor->PhysicalStart;
+          MemoryRegion[Index].Length = Hob.ResourceDescriptor->ResourceLength;
+        }
+
+        Index++;
+      }
+    }
+
+    Hob.Raw = GET_NEXT_HOB (Hob);
+  }
+
+  *MemoryRegionCount = Index;
+}
+
+/**
+  Function to compare 2 MM_IPL_MEMORY_REGION pointer based on Base.
+
+  @param[in] Buffer1            pointer to MM_IPL_MEMORY_REGION pointer to compare
+  @param[in] Buffer2            pointer to second MM_IPL_MEMORY_REGION pointer to compare
+
+  @retval 0                     Buffer1 equal to Buffer2
+  @retval <0                    Buffer1 is less than Buffer2
+  @retval >0                    Buffer1 is greater than Buffer2
+**/
+INTN
+EFIAPI
+MemoryRegionBaseAddressCompare (
+  IN CONST VOID  *Buffer1,
+  IN CONST VOID  *Buffer2
+  )
+{
+  if (((MM_IPL_MEMORY_REGION *)Buffer1)->Base > ((MM_IPL_MEMORY_REGION *)Buffer2)->Base) {
+    return 1;
+  } else if (((MM_IPL_MEMORY_REGION *)Buffer1)->Base < ((MM_IPL_MEMORY_REGION *)Buffer2)->Base) {
+    return -1;
+  }
+
+  return 0;
+}
+
+/**
+  Calculate the maximum support address.
+
+  @return the maximum support address.
+**/
+UINT8
+MmIplCalculateMaximumSupportAddress (
+  VOID
+  )
+{
+  UINT32  RegEax;
+  UINT8   PhysicalAddressBits;
+  VOID    *Hob;
+
+  //
+  // Get physical address bits supported.
+  //
+  Hob = GetFirstHob (EFI_HOB_TYPE_CPU);
+  if (Hob != NULL) {
+    PhysicalAddressBits = ((EFI_HOB_CPU *)Hob)->SizeOfMemorySpace;
+  } else {
+    AsmCpuid (0x80000000, &RegEax, NULL, NULL, NULL);
+    if (RegEax >= 0x80000008) {
+      AsmCpuid (0x80000008, &RegEax, NULL, NULL, NULL);

```

Comment:
do not use the hardcode value. Use definitions in Cpuid.h.

Reply:
yes, will update it